### PR TITLE
validation: checking if route already exists

### PR DESCRIPTION
**Pull request by jasmingacic**
Signed-off-by: jasmingacic <jasmin.gacic@gmail.com>

image

fixes https://github.com/kubeshop/kusk-gateway/issues/737
fixes https://github.com/kubeshop/kusk-gateway/issues/750

### CODE REVIEW

File: api/v1alpha1/api_webhook.go [hunk 1]
```diff
@@ -146,6 +149,39 @@ func (a *APIValidator) InjectDecoder(d *admission.Decoder) error {
...
+
+	// Get all API objects with this fleet field set
+	if err := a.Client.List(ctx, &apiObjs, &client.ListOptions{}); err != nil {
+		return fmt.Errorf("failure querying for the deployed APIs: %w", err)
+	}
+
+	// filter out apis are in the process of deletion
+	for _, api := range apiObjs.Items {
+		if api.Spec.Fleet.Name == fleet.Name && api.Spec.Fleet.Namespace == fleet.Namespace {
+			if api.ObjectMeta.DeletionTimestamp.IsZero() {
```
Comment (kylehodgetts) [inline api/v1alpha1/api_webhook.go:168]: you can combine this if statement into the if statement above

File: api/v1alpha1/api_webhook.go [hunk 2]
```diff
@@ -146,6 +149,39 @@ func (a *APIValidator) InjectDecoder(d *admission.Decoder) error {
...
+	}
+
+	// Get all API objects with this fleet field set
+	if err := a.Client.List(ctx, &apiObjs, &client.ListOptions{}); err != nil {
+		return fmt.Errorf("failure querying for the deployed APIs: %w", err)
+	}
+
+	// filter out apis are in the process of deletion
+	for _, api := range apiObjs.Items {
+		if api.Spec.Fleet.Name == fleet.Name && api.Spec.Fleet.Namespace == fleet.Namespace {
```
Comment (kylehodgetts) [inline api/v1alpha1/api_webhook.go:167]: You can invert this and turn it into a guard clause with a `continue`

Comment (jasmingacic) [reply api/v1alpha1/api_webhook.go:167]: I'll just combine the statement above

**Review by kylehodgetts — DISMISSED** [review]

**Comment by jasmingacic** [comment]
Added one more improvement. Listing all duplicate routes so users can get better message:
image

File: api/v1alpha1/api_webhook.go [hunk 3]
```diff
@@ -146,6 +151,42 @@ func (a *APIValidator) InjectDecoder(d *admission.Decoder) error {
...
+
+			for path := range newApiSpec.Paths {
+				if p, ok := apiSpec.Paths[path]; ok {
+					duplicates[path] = p
+				}
+			}
+		}
+	}
+	if len(duplicates) > 0 {
+		return fmt.Errorf("paths %s already exist with envoyfleet %s", reflect.ValueOf(duplicates).MapKeys(), fleet)
```
Comment (jasmingacic) [inline api/v1alpha1/api_webhook.go:184]: This message should allow users to debug what's wrong with their API

**Pull request merged**